pennylane/qchem/tapering.py

```diff
@@ -578,22 +579,30 @@ def _build_generator(operation, wire_order, op_gen=None):
     if op_gen is None:
         if operation.num_params < 1:  # Non-parameterized gates
             gen_mat = 1j * scipy.linalg.logm(qml.matrix(operation, wire_order=wire_order))
-            op_gen = qml.pauli_decompose(gen_mat, wire_order=wire_order, hide_identity=True)
-            op_gen = qml.simplify(op_gen)
-            if op_gen.ops[0].label() == qml.Identity(wires=[wire_order[0]]).label():
-                op_gen -= qml.Hamiltonian([op_gen.coeffs[0]], [qml.Identity(wires=wire_order[0])])
+            op_gen = qml.pauli_decompose(
+                gen_mat, wire_order=wire_order, hide_identity=True, pauli=True
+            )
+            op_gen.simplify()
+            op_gen.pop(PauliWord({}), 0.0)
         else:  # Single-parameter gates
             try:
-                op_gen = qml.generator(operation, "hamiltonian")
+                # TODO: simplify when qml.generator has a proper support for "arithmetic".
+                op_gen = (
+                    operation.generator()
+                    if active_new_opmath()
+                    else qml.generator(operation, "arithmetic")
+                ).pauli_rep
 
             except ValueError as exc:
                 raise NotImplementedError(
                     f"Generator for {operation} is not implemented, please provide it with 'op_gen' args."
                 ) from exc
     else:  # check that user-provided generator is correct
-        if not isinstance(op_gen, qml.Hamiltonian):
+        if not isinstance(op_gen, (qml.Hamiltonian, PauliSentence)) and not isinstance(
+            getattr(op_gen, "pauli_rep", None), PauliSentence
```

Comment:
if `pauli_rep` is either `None` or a `PauliSentence`, so `op_gen.pauli_rep is not None` should be sufficient.

Reply:
`qml.Hamiltonian` gives `None` as a `pauli_rep` so I decided to keep the condition different such that - check if it `Hamiltonian` or `PauliSentence` in the first condition and check if it is an arithmetic operator in the subsequent condition.